araig_gui/src/araig_gui/araig_gui.cpp

```diff
@@ -0,0 +1,230 @@
+  output_states_.resize(static_cast<std::vector<bool>::size_type>(num_outputs_));
+  spawnPubs();
+  spawnSubs();
+  stateInit();
+  ROS_INFO_STREAM("[GUI]: ARAIG_TEST_GUI initialized!");
+
+}
+
+void AraigGui::shutdownPlugin()
+{
+}
+
+void AraigGui::saveSettings(qt_gui_cpp::Settings& plugin_settings,
+    qt_gui_cpp::Settings& instance_settings) const
+{
+}
+
+void AraigGui::restoreSettings(const qt_gui_cpp::Settings& plugin_settings,
+    const qt_gui_cpp::Settings& instance_settings)
+{
+}
+
+void AraigGui::spawnPubs()
+{
+  for(int i=0; i<num_outputs_; i++)
+  {
+    std::string topicName = "/signal/ui/" + name_outputs_[static_cast<std::vector<std::string>::size_type>(i)];
+    output_pubs_[static_cast<std::vector<ros::Publisher>::size_type>(i)]
+        = nh_.advertise<araig_msgs::BoolStamped>(topicName, 10);
+    ROS_INFO_STREAM("[GUI]: Spawned output topic publisher " << i << " for : " << topicName);
+  }
+}
+
+void AraigGui::spawnSubs()
+{
+  for(int i=0; i<num_inputs_; i++)
+  {
+    std::string topicName = "/signal/runner/" + name_inputs_[static_cast<std::vector<std::string>::size_type>(i)];
+    input_subs_[static_cast<std::vector<ros::Subscriber>::size_type>(i)]
+        = nh_.subscribe<araig_msgs::BoolStamped>(topicName, 10, boost::bind(&AraigGui::callbackBool, this, _1, i));
+    ROS_INFO_STREAM("[GUI]: Spawned input topic subscriber " << i << " for : " << topicName);
+  }
+}
+
+void AraigGui::callbackBool(const araig_msgs::BoolStamped::ConstPtr &msg, const int &topicIdx)
+{
+  input_states_[static_cast<std::vector<bool>::size_type>(topicIdx)] = msg->data? true:false;
+  outputTestState();
+}
+
+void AraigGui::outputTestState()
+{
+  //input: 0: completed, 1: succeeded, 2:failed
+  completed_ = input_states_[0];
+  succ_ = input_states_[1];
+  failed_ = input_states_[2];
+
+  if(test_ready_) // test ready
+  {
+    ui_.lbTestState->setText("Test ready!");
+    ui_.lbTestResult->setText("Wait for start!");
+  }
+  else if(!completed_) // test running
+  {
+    ui_.lbTestState->setText("Test running!");
+    ui_.lbTestResult->setText("Waiting for result...");
+  }
+  else
+  {
+    if(interrupt_test_) // test interrupted
+    {
+      ui_.lbTestState->setText("Test interrupted!");
+      ui_.lbTestResult->setText("Test failed!");
+    }
+    else if(result_recorded_) // test recorded
+    {
+      ui_.lbTestState->setText("Please reset!");
+      ui_.lbTestResult->setText("Result recorded!");
+    }
+    else if(failed_ == succ_) // error in result
+    {
+      ui_.lbTestState->setText("Test completed!");
+      ui_.lbTestResult->setText("Error in result!");
+    }
+    else if(failed_) // test fails
+    {
+      ui_.lbTestState->setText("Test completed!");
+      ui_.lbTestResult->setText("Test failed!");
+    }
+    else if(succ_) // test succeeds
+    {
+      ui_.lbTestState->setText("Test completed!");
+      ui_.lbTestResult->setText("Test succeeded!");
+    }
+    else // otherwise
```

Comment:
Now this is the dangerous part of this state machine. If none of the above cases work, it means you have a weird undefined state and no matter what the problem is, you always see the same output "Please reset!", so it becomes challenging to identify what went wrong.